core/src/main/resources/schema/app/20180511133938_BUG-102745_add_gateway_topologies_table.sql

```diff
@@ -0,0 +1,31 @@
+-- // BUG-102745
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE gateway
+    ADD CONSTRAINT gateway_pkey PRIMARY KEY (id);
+
+CREATE TABLE gateway_topology (
+    id bigserial NOT NULL,
+    gateway_id int8 NOT NULL,
+    topologyname VARCHAR (255) NULL,
+    exposedservices text NULL,
+    CONSTRAINT gateway_topology_pkey PRIMARY KEY (id),
+    CONSTRAINT fk_gateway_gateway_topology FOREIGN KEY (gateway_id) REFERENCES gateway(id)
+);
+
+INSERT INTO gateway_topology(gateway_id, topologyname, exposedservices)
```

Comment:
Because we have canary release in production next to final release, after this change final version will still store data in the original table. Only canary will use the new structure, so records create during canary to final process (which is more than 2 hours) will be lost :( on this way

Reply:
Possible solution: The Gateway#getTopologies method should add the topology to the topology list from the topologyName and exposedServices fields if no topology is present in the topology list with that name.

Reply:
Like this:
```
public Set<GatewayTopology> getTopologies() {
    if (StringUtils.isNotEmpty(topologyName)) {
        if (topologies.stream().noneMatch(t -> t.getTopologyName().equals(topologyName))) {
            GatewayTopology gatewayTopology = new GatewayTopology();
            gatewayTopology.setTopologyName(topologyName);
            if (exposedServices != null && StringUtils.isNoneEmpty(exposedServices.getValue())) {
                gatewayTopology.setExposedServices(exposedServices);
            }
            topologies.add(gatewayTopology);
        }
    }
    return topologies;
}
```